### PR TITLE
Do not discern between statements with and without semicolon after lowering to HIR

### PR DESCRIPTION
**Pull request by petrochenkov**
In AST we need to discern between statements with semicolon and statements without semicolon in a block to detect the block's "trailing expression" (which may end up trailing after macro expansion, but start somewhere in the middle of the block, see the in-progress issue https://github.com/rust-lang/rust/issues/61733).

After the trailing expression is detected during lowering to HIR, the distinction is no longer necessary...

... with exception of one surprising corner case that I discovered today:
```rust
fn main() {
    0; // OK, unused expression

    ()
}
fn main() {
    { 0 }; // OK, unused expression

    ()
}
fn main() {
    { 0 } // ERROR expected (), found integer (???)

    ()
}
```
Apparently semicolon-less expression in the middle of a block must unify with `()` for some reason.

The reason is unclear to me, so this PR removes this special case and unifies `StmtKind::Expr` and `StmtKind::Semi` in HIR.
r? @nikomatsakis

### CODE REVIEW

**Comment by petrochenkov** [comment]
cc @rust-lang/lang

File: src/librustc/hir/mod.rs [hunk 1]
```diff
@@ -1205,20 +1205,16 @@ pub enum StmtKind {
     /// An item binding.
     Item(ItemId),
 
-    /// An expression without a trailing semi-colon (must have unit type).
+    /// An expression statement.
```
Comment (Centril) [inline src/librustc/hir/mod.rs]:
```suggestion
    /// An expression statement (may have any type).
```

File: src/librustc/hir/lowering.rs [hunk 2]
```diff
@@ -4877,12 +4877,11 @@ impl<'a> LoweringContext<'a> {
...
                 let body_block = self.with_loop_scope(e.id, |this| this.lower_block(body, false));
                 let body_expr = P(self.expr_block(body_block, ThinVec::new()));
-                let body_stmt = self.stmt(body.span, hir::StmtKind::Expr(body_expr));
 
                 let loop_block = P(self.block_all(
                     e.span,
-                    hir_vec![next_let, match_stmt, pat_let, body_stmt],
-                    None,
+                    hir_vec![next_let, match_stmt, pat_let],
+                    Some(body_expr),
```
Comment (Centril) [inline src/librustc/hir/lowering.rs]: Can you update the comment that outlines the desugaring at the top of the match arm?

File: src/librustc_typeck/check/mod.rs [hunk 3]
```diff
@@ -4856,10 +4856,6 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             // Ignore for now.
             hir::StmtKind::Item(_) => {}
             hir::StmtKind::Expr(ref expr) => {
-                // Check with expected type of `()`.
-                self.check_expr_has_type_or_error(&expr, self.tcx.mk_unit());
```
Comment (Centril) [inline src/librustc_typeck/check/mod.rs]: (T-Lang: N.B. this removes the expectation that `{ expr } : ()` and so therefore:
```rust
{ core::default::Default::default() }
1;
```
may stop compiling (crater will find out).

We could add inference defaulting to `()` if that is desirable.

**Comment by cramertj** [comment]
> Apparently semicolon-less expression in the middle of a block must unify with () for some reason.

Huh, that's the behavior I would've expected-- my intuition was that intermediate statements should always be of type `()`, and that using `;` turns a `T`-returning-expression into a `()`-returning expression.

**Comment by eddyb** [comment]
Yeah, the intention is that `;` throws away the value but otherwise every statement must evaluate to `()`. I suspect this PR might fail on crater due removing information from inference .

**Comment by petrochenkov** [comment]
@craterbot run mode=check-only

**Comment by petrochenkov** [comment]
@cramertj @eddyb 
My intuition was that
- statements don't have types or values at all
- results of expressions in trailing position are returned, so their type is restricted by the return type
- results of expressions in the middle of a block are thrown away, so their type is not restricted

**Comment by petrochenkov** [comment]
Hmm, all the errors are in the "does not live long enough" category, rather than regressed type inference.

This is due to changes in `for` loop desugaring.
Previously `for` loops implicitly added a semicolon to the end of their body.
```rust
for x in xs {
    doit()
}

=>

for x in xs {
    doit();
}
```
(Not other block or loop constructions do that.)

I'll revert this part and you'll see what effect it has on tests.

File: src/test/ui/mismatched_types/for-loop-has-unit-body.rs [hunk 4]
```diff
@@ -1,5 +1,7 @@
+// compile-pass
+
 fn main() {
     for x in 0..3 {
-        x //~ ERROR mismatched types
+        x
```
Comment (petrochenkov) [inline src/test/ui/mismatched_types/for-loop-has-unit-body.rs]: `x` no longer needs to unify with `()`, so it's no longer an error.

File: src/test/ui/for/for-loop-has-unit-body.rs [hunk 5]
```diff
@@ -1,8 +1,7 @@
-// run-pass
 fn main() {
     // Check that the tail statement in the body unifies with something
     for _ in 0..3 {
-        unsafe { std::mem::uninitialized() }
+        unsafe { std::mem::uninitialized() } //~ ERROR type annotations needed
```
Comment (petrochenkov) [inline src/test/ui/for/for-loop-has-unit-body.rs]: `unsafe { std::mem::uninitialized() }` no longer needs to unify with `()`, so it cannot infer its type.

**Comment by petrochenkov** [comment]
Updated with `for` loop desugaring changes reverted.

**Comment by eddyb** [comment]
@petrochenkov Can you open an issue about the `for` inconsistency?

**Comment by petrochenkov** [comment]
@eddyb 
>Can you open an issue about the for inconsistency?

https://github.com/rust-lang/rust/issues/61902

**Comment by cramertj** [comment]
We discussed this in the @rust-lang/lang meeting and we agreed that the current behavior is roughly as-intended (implementation inconsistencies aside), and that this change and similar ones should go through the RFC process.

**Comment by petrochenkov** [comment]
Ok, this is not something important enough to go through an RFC.
A documentation issue for this quirk - https://github.com/rust-lang-nursery/reference/issues/625.

The `for` loop inconsistency (https://github.com/rust-lang/rust/issues/61902) still needs some attention.

**Pull request closed without merging**